### ISSUE TITLE
Question about the Implementation of Wasserstein Loss

### ISSUE THREAD

**Issue by jueqiw** [xiat0616/BrainAgeing#1]
Dear Dr. Tian Xia,

Thank you so much for this wonderful work on brain ageing. I'm certain this will be of ongoing value to the research community as a whole. I am also working on predicting future brain MRIs, so reading your code is really helpful to me.

I have a question about the [Wasserstein loss](https://github.com/xiat0616/BrainAgeing/blob/5eca207fe4815c93fba6ccdbec2266387ef21255/utils/cost.py#L7) in your code. Why it is `K.mean(y_true * y_pred)`? I thought it should be `loss = discriminator(fake) - discriminator(real)` according to the WGAN paper.

I would very much appreciate it if you could point me in the right direction.

Thanks.

**Comment by xiat0616** [comment]
Hi,
Thank you for your interest in this project.

This is because of the limitation of the keras package. Your understanding of Wassersten loss is correct. But here I implemented in the way where y_pred is either 1 or 0. You can find in the executor code I set this as

real_pred = -np.ones((old_img.shape[0],1))
fake_pred = np.ones((old_img.shape[0],1))

Eventually, the training is using the Wasserstein loss as you understood.

Best,
Tian

**Comment by jueqiw** [comment]
Hi Tian,

Thanks for answering my question! Sorry for that I am not familiar with Keras package. I understand now. So because of this limitation, you are creating the Wasserstein loss by:

```py
fake_pred * discriminator(fake) + real_pred * discriminator(real)
= 1 * discriminator(fake) - 1 * discriminator(real)
```

Thanks! I will close this issue.

Jueqi
